src/DataStructures/Tensor/Expressions/Evaluate.hpp

```diff
@@ -38,8 +40,157 @@ constexpr bool contains_indices_to_contract(
+/// that is being checked, e.g. the type of `ti_a`
+template <typename LhsIndexList, typename RhsIndexList,
+          typename LhsTensorIndexList, typename RhsTensorIndexList,
+          typename CurrentLhsTensorIndex>
+struct EvaluateIndexCheckHelper {
+  using lhs_index =
+      tmpl::at<LhsIndexList,
+               tmpl::index_of<LhsTensorIndexList, CurrentLhsTensorIndex>>;
+  using rhs_index =
+      tmpl::at<RhsIndexList,
+               tmpl::index_of<RhsTensorIndexList, CurrentLhsTensorIndex>>;
+
+  using type = std::bool_constant<
+      lhs_index::ul == rhs_index::ul and
+      std::is_same_v<typename lhs_index::Frame, typename rhs_index::Frame> and
+      ((lhs_index::index_type == rhs_index::index_type and
+        lhs_index::dim == rhs_index::dim) or
+       (lhs_index::index_type == IndexType::Spacetime and
+        lhs_index::dim == rhs_index::dim + 1) or
+       (rhs_index::index_type == IndexType::Spacetime and
+        lhs_index::dim + 1 == rhs_index::dim))>;
+};
+
+/// \brief Check that a RHS tensor's indices can be evaluated to their
+/// corresponding indices in the LHS tensor
+///
+/// \details
+/// For more details, see `EvaluateIndexCheckHelper`, which performs the check
+/// for each index one at a time.
+///
+/// \tparam LhsIndexList the LHS tensor's \ref SpacetimeIndex "TensorIndexType"
+/// list
+/// \tparam RhsIndexList the RHS tensor's \ref SpacetimeIndex "TensorIndexType"
+/// list
+/// \tparam LhsTensorIndexList the LHS tensor's generic index list
+/// \tparam RhsTensorIndexList the RHS tensor's generic index list
+template <typename LhsIndexList, typename RhsIndexList,
+          typename LhsTensorIndexList, typename RhsTensorIndexList>
+using EvaluateIndexCheck =
+    tmpl::fold<LhsTensorIndexList, tmpl::bool_<true>,
+               tmpl::and_<tmpl::_state,
+                          EvaluateIndexCheckHelper<
+                              tmpl::pin<LhsIndexList>, tmpl::pin<RhsIndexList>,
+                              tmpl::pin<LhsTensorIndexList>,
+                              tmpl::pin<RhsTensorIndexList>, tmpl::_element>>>;
 }  // namespace detail
 
+/// \brief Computes a transformation from the LHS tensor's multi-indices to
+/// the equivalent RHS tensor's multi-indices, according to the differences in
+/// the orderings of their generic indices
+///
+/// \details
+/// The elements of the transformation are the positions of the RHS generic
+/// indices in the LHS generic indices. Put another way, for some `i`,
+/// `rhs_tensorindices[i] == lhs_tensorindices[index_transformation[i]]`.
+///
+/// Here is an example of what the algorithm does:
+///
+/// Tensor equation: \f$L_{cab} = R_{abc}\f$
+/// `lhs_tensorindices`:
+/// \code
+/// {2, 0, 1} // i.e. {c, a, b}
+/// \endcode
+/// `rhs_tensorindices`:
+/// \code
+/// {0, 1, 2} // i.e. {a, b, c}
+/// \endcode
+/// returned `index_transformation`:
+/// \code
+/// {1, 2, 0} // positions of RHS indices {c, a, b} in LHS indices {a, b, c}
```

Comment:
Yes, you're correct. I didn't notice the mistake because I just copy + pasted this from `TensorAsExpression.hpp`, so the mistake must not have been caught earlier by myself and during review. I'll correct this one in `Evaluate.hpp` in a fixup for this PR, but the one I copy + pasted from in `TensorAsExpression.hpp` still has this mistake. I have a PR coming down the pipe after this PR to factor this function out of both files. Let me know which you would prefer:
(1) Also correct the example in `TensorAsExpression.hpp` as a separate commit in this PR
(2) Leave the incorrect wording in `TensorAsExpression.hpp` for now, but make sure that when I factor this function out of `Evaluate.hpp` and `TensorAsExpression.hpp` in the next PR, I use the corrected wording (the wording I will use in a fixup for `Evaluate.hpp`  on this PR)

Reply:
If (1) is as straightforward as it sounds (a small commit with the wording fix), then I'd say let's do (1) and get it cleared up right away. But if that'll make your work annoying in any way, I'm happy with (2). From here, you decide 🙂

Reply:
(1) is good with me! I'll push the fixes soon

Reply:
And yes, it's just a wording fix. No change to the actual code

Reply:
@macedo22 feel free to squash directly (this + other changes)

Reply:
Would you still like the separate wording fix to `TensorAsExpression.hpp` as its own new 4th commit?

Reply:
Yes, that sounds good 👍

Reply:
Oops, scratch that. I reread previous comments, and it sounds like "yes"